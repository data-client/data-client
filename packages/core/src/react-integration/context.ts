import { createContext } from 'react';
import type { DenormalizeCache } from '@rest-hooks/normalizr';
import type { ActionTypes } from '@rest-hooks/core/types';
import { initialState } from '@rest-hooks/core/state/reducer';
import Controller from '@rest-hooks/core/controller/Controller';

export const StateContext = createContext(initialState);

const dispatch = (value: ActionTypes) => {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    console.error(
      'It appears you are trying to use Rest Hooks without a provider.\nFollow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component',
    );
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'test') {
      console.error(
        'If you are trying to test: https://resthooks.io/docs/guides/unit-testing-hooks',
      );
    }
  }
  return Promise.resolve();
};
export const DispatchContext = createContext(dispatch);

// this is not needed anymore, but keeping around for backcompatibility
export const DenormalizeCacheContext = createContext<DenormalizeCache>({
  entities: {},
  results: {},
});

export const ControllerContext = createContext<Controller>(
  new Controller({
    dispatch,
    globalCache: {
      entities: {},
      results: {},
    },
  }),
);
