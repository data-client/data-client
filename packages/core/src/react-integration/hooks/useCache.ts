import { ReadShape, ParamsFromShape } from '@rest-hooks/core/endpoint/index';
import { DenormalizeNullable } from '@rest-hooks/endpoint';
import { useDenormalized } from '@rest-hooks/core/state/selectors/index';
import { useContext, useMemo } from 'react';
import { StateContext } from '@rest-hooks/core/react-integration/context';
import { denormalize, inferResults } from '@rest-hooks/normalizr';
import { ExpiryStatus } from '@rest-hooks/core/controller/Expiry';

/**
 * Access a resource if it is available.
 *
 * `useCache` guarantees referential equality globally.
 * @see https://resthooks.io/docs/api/useCache
 */
export default function useCache<
  Shape extends Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
>(
  fetchShape: Shape,
  params: ParamsFromShape<Shape> | null,
): DenormalizeNullable<Shape['schema']> {
  const state = useContext(StateContext);

  const { data, expiryStatus, expiresAt } = useDenormalized(
    fetchShape,
    params,
    state,
  );
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  /*********** This block is to ensure results are only filled when they would not suspend **************/
  // This computation reflects the behavior of useResource/useRetrive
  // It only changes the value when expiry or params change.
  // This way, random unrelated re-renders don't cause the concept of expiry
  // to change
  const expired = useMemo(() => {
    if ((Date.now() <= expiresAt && !forceFetch) || !params) return false;
    return true;
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, params && fetchShape.getFetchKey(params), forceFetch]);

  // if useResource() would suspend, don't include entities from cache
  if (expiryStatus !== ExpiryStatus.Valid && expired) {
    return denormalize(
      inferResults(fetchShape.schema, [params], state.indexes),
      fetchShape.schema,
      {},
    )[0];
  }
  /*********************** end block *****************************/

  return data;
}
