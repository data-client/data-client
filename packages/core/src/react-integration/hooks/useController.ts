import { useContext } from 'react';
import Controller from '@rest-hooks/core/controller/Controller';
import { ControllerContext } from '@rest-hooks/core/react-integration/context';

/**
 * Imperative control of Rest Hooks store
 * @see https://resthooks.io/docs/api/useController
 */
export default function useController(): Controller {
  return useContext(ControllerContext);
}
