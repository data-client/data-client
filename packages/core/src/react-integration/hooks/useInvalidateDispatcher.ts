import { useContext, useCallback } from 'react';
import { ReadShape, ParamsFromShape } from '@rest-hooks/core/endpoint/index';
import { DispatchContext } from '@rest-hooks/core/react-integration/context';
import { INVALIDATE_TYPE } from '@rest-hooks/core/actionTypes';

/** Invalidate a certain item within the cache
 * @deprecated use https://resthooks.io/docs/api/Controller#invalidate
 */
export default function useInvalidateDispatcher(): <
  Shape extends ReadShape<any, any>,
>(
  fetchShape: Shape,
  params: ParamsFromShape<Shape>,
) => void {
  const dispatch = useContext(DispatchContext);

  const invalidateDispatcher = useCallback(
    <Shape extends ReadShape<any, any>>(
      fetchShape: Shape,
      params: ParamsFromShape<Shape>,
    ) => {
      dispatch({
        type: INVALIDATE_TYPE,
        meta: {
          key: fetchShape.getFetchKey(params),
        },
      });
    },
    [dispatch],
  );

  return invalidateDispatcher;
}
