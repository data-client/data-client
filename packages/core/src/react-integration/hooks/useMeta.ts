import { useContext, useMemo } from 'react';
import { EndpointInterface, FetchFunction } from '@rest-hooks/endpoint';

import { FetchShape, ParamsFromShape } from '../../endpoint/index.js';
import { StateContext } from '../context.js';
import { selectMeta } from '../../state/selectors/index.js';

/**
 * Gets meta for a fetch key.
 * @see https://resthooks.io/docs/api/useMeta
 */
export default function useMeta<
  E extends
    | Pick<EndpointInterface<FetchFunction>, 'key'>
    | Pick<FetchShape<any, any>, 'getFetchKey'>,
  Args extends
    | (E extends { key: any }
        ? readonly [...Parameters<E['key']>]
        : readonly [ParamsFromShape<E>])
    | readonly [null],
>(endpoint: E, ...args: Args) {
  const state = useContext(StateContext);
  const key = args[0]
    ? (endpoint as any).key
      ? (endpoint as any).key(...args)
      : (endpoint as any).getFetchKey(args[0])
    : '';

  return useMemo(() => {
    if (!key) return null;
    return selectMeta(state, key);
  }, [key, state]);
}
