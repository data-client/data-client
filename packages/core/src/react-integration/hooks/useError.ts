import { ReadShape, ParamsFromShape } from '@rest-hooks/core/endpoint/index';
import { NetworkError, UnknownError } from '@rest-hooks/core/types';
import { StateContext } from '@rest-hooks/core/react-integration/context';
import { useContext, useMemo } from 'react';
import useController from '@rest-hooks/core/react-integration/hooks/useController';
import shapeToEndpoint from '@rest-hooks/core/endpoint/adapter';

export type ErrorTypes = NetworkError | UnknownError;

type UseErrorReturn<P> = P extends null ? undefined : ErrorTypes | undefined;

/**
 * Get any errors for a given request
 * @see https://resthooks.io/docs/api/useError
 */
export default function useError<
  Shape extends Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
>(
  fetchShape: Shape,
  params: ParamsFromShape<Shape> | null,
): UseErrorReturn<typeof params> {
  const state = useContext(StateContext);

  const controller = useController();

  const endpoint = useMemo(() => {
    return shapeToEndpoint(fetchShape);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return controller.getError(endpoint, params, state) as any;
}
