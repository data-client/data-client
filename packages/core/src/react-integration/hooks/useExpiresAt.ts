import { ReadShape, ParamsFromShape } from '@rest-hooks/core/endpoint/index';
import useMeta from '@rest-hooks/core/react-integration/hooks/useMeta';

/** Returns whether the data at this url is fresh or stale
 * @deprecated use https://resthooks.io/docs/api/Controller#getResponse
 */
export default function useExpiresAt<
  Shape extends Pick<ReadShape<any, any>, 'getFetchKey' | 'options'>,
>(
  fetchShape: Shape,
  params: ParamsFromShape<Shape> | null,
  entitiesExpireAt = 0,
): number {
  const meta = useMeta(fetchShape, params);

  if (!meta) {
    return entitiesExpireAt;
  }

  return meta.expiresAt;
}
