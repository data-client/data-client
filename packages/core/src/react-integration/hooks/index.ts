import useFetcher from '@rest-hooks/core/react-integration/hooks/useFetcher';
import useController from '@rest-hooks/core/react-integration/hooks/useController';
import useCache from '@rest-hooks/core/react-integration/hooks/useCache';
import useRetrieve from '@rest-hooks/core/react-integration/hooks/useRetrieve';
import useResource from '@rest-hooks/core/react-integration/hooks/useResource';
import useSubscription from '@rest-hooks/core/react-integration/hooks/useSubscription';
import useMeta from '@rest-hooks/core/react-integration/hooks/useMeta';
import useError from '@rest-hooks/core/react-integration/hooks/useError';
import useExpiresAt from '@rest-hooks/core/react-integration/hooks/useExpiresAt';
import useInvalidator from '@rest-hooks/core/react-integration/hooks/useInvalidator';
import useResetter from '@rest-hooks/core/react-integration/hooks/useResetter';
import useFetchDispatcher from '@rest-hooks/core/react-integration/hooks/useFetchDispatcher';
import useInvalidateDispatcher from '@rest-hooks/core/react-integration/hooks/useInvalidateDispatcher';
import useFetch from '@rest-hooks/core/react-integration/newhooks/useFetch';
import useSuspense from '@rest-hooks/core/react-integration/newhooks/useSuspense';
export { default as hasUsableData } from '@rest-hooks/core/react-integration/hooks/hasUsableData';
export type { ErrorTypes } from '@rest-hooks/core/react-integration/hooks/useError';

export {
  useFetcher,
  useController,
  useFetchDispatcher,
  useCache,
  useError,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useExpiresAt,
  useInvalidator,
  useInvalidateDispatcher,
  useResetter,
  useFetch,
  useSuspense,
};
