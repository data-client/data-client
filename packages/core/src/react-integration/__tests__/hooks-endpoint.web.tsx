import { CoolerArticleResource, PaginatedArticleResource } from '__tests__/new';
import React, { Suspense, useEffect } from 'react';
import { render, act } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import nock from 'nock';

// relative imports to avoid circular dependency in tsconfig references
import {
  makeRenderRestHook,
  makeCacheProvider,
  mockInitialState,
} from '../../../../test';
import { ControllerContext, DispatchContext, StateContext } from '../context';
import { useController } from '../hooks';
import { State, ActionTypes } from '../../types';
import { INVALIDATE_TYPE, RESET_TYPE } from '../../actionTypes';
import { articlesPages } from '../test-fixtures';
import { Controller } from '../..';

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const controller = new Controller({ dispatch });
  const tree = (
    <ControllerContext.Provider value={controller}>
      <DispatchContext.Provider value={dispatch}>
        <Suspense fallback={null}>
          <Component />
        </Suspense>
      </DispatchContext.Provider>
    </ControllerContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalledTimes(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    delete call[0]?.meta?.createdAt;
    delete call[0]?.meta?.promise;
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function testRestHook(
  callback: () => void,
  state: State<unknown>,
  dispatch = (v: ActionTypes) => Promise.resolve(),
) {
  const controller = new Controller({ dispatch });
  return renderHook(callback, {
    wrapper: function Wrapper({ children }: { children: React.ReactNode }) {
      return (
        <ControllerContext.Provider value={controller}>
          <DispatchContext.Provider value={dispatch}>
            <StateContext.Provider value={state}>
              {children}
            </StateContext.Provider>
          </DispatchContext.Provider>
        </ControllerContext.Provider>
      );
    },
  });
}

let mynock: nock.Scope;

beforeAll(() => {
  nock(/.*/)
    .persist()
    .defaultReplyHeaders({
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    })
    .options(/.*/)
    .reply(200);
  mynock = nock(/.*/).defaultReplyHeaders({
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json',
  });
});

afterAll(() => {
  nock.cleanAll();
});

describe('useController.fetch', () => {
  const payload = { id: 1, content: 'hi' };

  it('should dispatch an action that fetches a create', async () => {
    mynock.post(`/article-cooler/`).reply(201, payload);

    function DispatchTester() {
      const { fetch } = useController();
      fetch(CoolerArticleResource.create(), { content: 'hi' }).then(v => {
        v.author;
        /*
        // @ts-expect-error
        v.doesnotexist;*/
      });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should handle zero argument Endpoints', async () => {
    const endpoint = CoolerArticleResource.list().extend({
      fetch() {
        return CoolerArticleResource.list().call(this as any, {});
      },
      url() {
        return '';
      },
    });
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    function DispatchTester() {
      const { fetch } = useController();
      fetch(endpoint).then(v => {
        v[0].author;
        /*
        // @ts-expect-error
        v.doesnotexist;*/
      });
      return null;
    }
  });

  it('should dispatch an action with updater in the meta if update shapes params are passed in', async () => {
    mynock.post(`/article-cooler/`).reply(201, payload);

    function DispatchTester() {
      const { fetch } = useController();
      const params = { content: 'hi' };
      fetch(CoolerArticleResource.create(), params).then(v => {
        v.title;
        /*
        // @ts-expect-error
        v.doesnotexist;*/
      });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should refresh get details', async () => {
    mynock.get(`/article-cooler/${payload.id}`).reply(200, payload);

    function DispatchTester() {
      const { fetch } = useController();
      // @ts-expect-error
      () => fetch(CoolerArticleResource.detail(), {}, { content: 'hi' });
      fetch(CoolerArticleResource.detail(), { id: payload.id }).then(v => {
        v.title;
        /*
        // @ts-expect-error
        v.doesnotexist;*/
      });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should console.error() a warning when fetching without a Provider', () => {
    const oldError = console.error;
    const spy = (console.error = jest.fn());
    renderHook(() => {
      const { fetch } = useController();
      fetch(CoolerArticleResource.create(), { content: 'hi' });
      return null;
    });
    expect(spy.mock.calls[0]).toMatchInlineSnapshot(`
      Array [
        "It appears you are trying to use Rest Hooks without a provider.
      Follow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component",
      ]
    `);
    console.error = oldError;
  });

  it('should dispatch an action that fetches a partial update', async () => {
    mynock.patch(`/article-cooler/1`).reply(200, payload);

    function DispatchTester() {
      const { fetch } = useController();
      fetch(
        CoolerArticleResource.partialUpdate(),
        { id: payload.id },
        { content: 'changed' },
      );
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should dispatch an action that fetches a full update', async () => {
    mynock.put(`/article-cooler/1`).reply(200, payload);

    function DispatchTester() {
      const { fetch } = useController();
      fetch(
        CoolerArticleResource.update(),
        { id: payload.id },
        { content: 'changed' },
      );
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });
});

describe('useController().invalidate', () => {
  it('should not invalidate anything if params is null', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.list(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let invalidate: any;
    testRestHook(
      () => {
        invalidate = useController().invalidate;
      },
      state,
      dispatch,
    );
    invalidate(PaginatedArticleResource.list(), null);
    expect(dispatch).not.toHaveBeenCalled();
  });
  it('should return a function that dispatches an action to invalidate a resource', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.list(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let invalidate: any;
    testRestHook(
      () => {
        invalidate = useController().invalidate;
      },
      state,
      dispatch,
    );
    invalidate(PaginatedArticleResource.list(), {});
    expect(dispatch).toHaveBeenCalledWith({
      type: INVALIDATE_TYPE,
      meta: {
        key: 'GET http://test.com/article-paginated/',
      },
    });
  });
  it('should return the same === function each time', () => {
    const track = jest.fn();

    const { rerender } = renderHook(() => {
      const { invalidate } = useController();
      useEffect(track, [invalidate]);
    });
    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 4; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});

describe('useController().reset', () => {
  afterEach(() => {
    jest.useRealTimers();
  });
  it('should return a function that dispatches an action to reset the cache', async () => {
    jest.useFakeTimers();
    const state = mockInitialState([
      {
        endpoint: PaginatedArticleResource.list(),
        args: [{}],
        response: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let reset: () => Promise<void> = () => Promise.resolve();
    testRestHook(
      () => {
        reset = useController().resetEntireStore;
      },
      state,
      dispatch,
    );
    await act(reset);
    expect(dispatch).toHaveBeenCalledWith({
      type: RESET_TYPE,
      date: Date.now(),
    });
  });
  it('should return the same === function each time', () => {
    const track = jest.fn();

    const { rerender } = renderHook(() => {
      const reset = useController().resetEntireStore;
      useEffect(track, [reset]);
    });
    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 4; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});
