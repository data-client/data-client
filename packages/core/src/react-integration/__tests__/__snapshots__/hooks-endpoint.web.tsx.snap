// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useController.fetch should dispatch an action that fetches a create 1`] = `
Object {
  "endpoint": [Function],
  "meta": Object {
    "args": Array [
      Object {
        "content": "hi",
      },
    ],
    "key": "POST http://test.com/article-cooler/",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useController.fetch should dispatch an action that fetches a full update 1`] = `
Object {
  "endpoint": [Function],
  "meta": Object {
    "args": Array [
      Object {
        "id": 1,
      },
      Object {
        "content": "changed",
      },
    ],
    "key": "PUT http://test.com/article-cooler/1",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useController.fetch should dispatch an action that fetches a partial update 1`] = `
Object {
  "endpoint": [Function],
  "meta": Object {
    "args": Array [
      Object {
        "id": 1,
      },
      Object {
        "content": "changed",
      },
    ],
    "key": "PATCH http://test.com/article-cooler/1",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useController.fetch should dispatch an action with updater in the meta if update shapes params are passed in 1`] = `
Object {
  "endpoint": [Function],
  "meta": Object {
    "args": Array [
      Object {
        "content": "hi",
      },
    ],
    "key": "POST http://test.com/article-cooler/",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useController.fetch should refresh get details 1`] = `
Object {
  "endpoint": [Function],
  "meta": Object {
    "args": Array [
      Object {
        "id": 1,
      },
    ],
    "key": "GET http://test.com/article-cooler/1",
    "options": [Function],
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;
